stix2/environment.py

```diff
@@ -186,3 +192,306 @@ def creator_of(self, obj):
             return self.get(creator_id)
         else:
             return None
+
+    @staticmethod
+    def semantically_equivalent(obj1, obj2, **weight_dict):
+        """This method is meant to verify if two objects of the same type are
+        semantically equivalent.
+
+        Args:
+            obj1: A stix2 object instance
+            obj2: A stix2 object instance
+            weight_dict: A dictionary that can be used to override settings
+                in the semantic equivalence process
+
+        Returns:
+            float: A number between 0.0 and 100.0 as a measurement of equivalence.
+
+        Warnings:
+            Not all objects are supported.
+
+        Notes:
+            This implementation follows the Committee Note on semantic equivalence.
+            see `the Committee Note <link here>`__.
+
+        """
+        # default weights used for the semantic equivalence process
+        weigths = {
+            "attack-pattern": {
+                "name": 30,
+                "external_references": 70,
+            },
+            "campaign": {
+                "name": 60,
+                "aliases": 40,
+            },
+            "identity": {
+                "name": 60,
+                "identity_class": 20,
+                "sectors": 20,
+            },
+            "indicator": {
+                "indicator_types": 15,
+                "pattern": 80,
+                "valid_from": 5,
+            },
+            "location": {
+                "longitude_latitude": 34,
+                "region": 33,
+                "country": 33,
+            },
+            "malware": {
+                "malware_types": 20,
+                "name": 80,
+            },
+            "threat-actor": {
+                "name": 60,
+                "threat_actor_types": 20,
+                "aliases": 20,
+            },
+            "tool": {
+                "tool_types": 20,
+                "name": 80,
+            },
+            "vulnerability": {
+                "name": 30,
+                "external_references": 70,
+            },
+            "_internal": {
+                "tdelta": 1,  # One day interval
+            },
+        }
+
+        if weight_dict:
+            weigths.update(weight_dict)
+
+        matching_score = 0.0
+        sum_weights = 0.0
+        type1, type2 = obj1["type"], obj2["type"]
+
+        if type1 != type2:
+            raise ValueError('The objects to compare must be of the same type!')
+
+        if obj1.get("spec_version", "2.0") != obj2.get("spec_version", "2.0"):
+            raise ValueError('The objects to compare must be of the same spec version!')
```

Comment:
There might be cases where it makes sense to compare across spec versions. Could we only raise this error if there's a property on one object but not on the other that is used by the algorithm from the whitepaper? I think the only properties where this would be the case are the *_types properties. Down the road we might also look into tracking the properties that got renamed, so a 2.0 indicator with a label that matches a 2.1 indicator's indicator_type could be found to be equivalent.

Reply:
The Note does mention anything about cross-spec comparison or how we could figure out weights for these types of checks. What I could do is have the ability to enable or disable this check and if someone wants to write that kind of logic they could do so without having to modify the main logic.

Reply:
If I remember right the Note doesn't mention spec version at all. That approach of making it optional sounds good to me.

Reply:
Right. I meant to say "The Note does **not** mention..."